functions/gsheets.py

```diff
@@ -0,0 +1,238 @@
+        """
+        Load credentials file from SSM parameter.
+        """
+        try:
+            r = self.ssm.get_parameter(
+                Name=self.ssm_path,
+                WithDecryption=True,
+            )
+            data = r["Parameter"]["Value"]
+            with open(self.default_filename, "w") as f:
+                f.write(data)
+            self.creds = service_account.Credentials.from_service_account_file(
+                self.default_filename
+            )
+        except self.ssm.exceptions.ParameterNotFound:
+            logger.error(f"SSM Parameter {self.ssm_path} not found")
+            self.creds = None
+
+    def save_to_ssm(self, filename):
+        """
+        Save credentials file to SSM parameter.
+        """
+        if not filename:
+            filename = self.default_filename
+        if not exists(filename):
+            raise Exception(f"Credentials file {filename} not found")
+
+        data = ""
+        with open(filename, "r") as f:
+            data = json.dumps(f.read())
+
+        r = self.ssm.put_parameter(
+            Name=self.ssm_path,
+            Description="Service account creds for gsheets authorization",
+            Value=json.loads(data),
+            Type="SecureString",
+            Overwrite=True,
+        )
+        logger.info(f"Saved token version {r['Version']}")
+
+    def delete_from_ssm(self):
+        """
+        Delete credentials from SSM.
+        """
+        r = self.ssm.delete_parameter(Name=self.ssm_path)
+        if (r["ResponseMetadata"]["HTTPStatusCode"] == 200):
+            logger.info(f"Successfully destroyed ssm parameter {self.ssm_path}")
+        else:
+            logger.info(f"Failed to destroy ssm parameter {self.ssm_path}")
+
+
+class GSheet:
+    def __init__(self, spreadsheet_id, in_lambda=False):
+        self.spreadsheet_id = spreadsheet_id
+        auth = GSheetsAuth()
+        auth.load_from_ssm()
+        self.creds = auth.credentials
+
+    @property
+    def service(self):
+        return build(
+            "sheets", "v4", credentials=self.creds, cache_discovery=False
+        )
+
+    def _download_csv(self, sheet_name):
+        result = (
+            self.service.spreadsheets()
+            .values()
+            .get(spreadsheetId=self.spreadsheet_id, range=sheet_name)
+            .execute()
+        )
+
+        # /tmp is the only directory you can write to in a lambda function
+        file_path = f"/tmp/{sheet_name.replace('/', '-')}.csv"
+
+        with open(file_path, "w") as f:
+            writer = csv.writer(f)
+            writer.writerows(result.get("values"))
+
+        f.close()
+
+        logger.debug(f"Successfully downloaded {sheet_name}.csv")
+        return file_path
+
+    def import_to_dynamo(self, sheet_name):
+        file_path = self._download_csv(sheet_name)
+
+        schedule_csv_to_dynamo(SCHEDULE_TABLE, file_path)
+
+
+def schedule_json_to_dynamo(schedule_table, json_file=None, schedule_data=None):
```

Comment:
not asking for changes; just thinking out loud. i wrote this code iirc and it feels a bit inelegant and possibly error prone. I was fine with that when it only ran locally and i could play with it and comment out stuff. Now that it's going to run in a lambda that has to be packaged, deployed, etc, I wonder if we're going to wish it was broken up a bit and more testable. IDK, what do you think?

Reply:
Added a unit test for the parsing code - should add more as well.